### PR TITLE
fix: do not report an error with deprecated keyword

### LINKED ISSUE #714: Deprecated parameters make the server always return 422 response
**Describe the bug**
I cannot get a successful response if my spec has a deprecated parameter.

**To Reproduce**

1. Given this OpenAPI document:
```openapi: 3.0.1
info:
  title: Test
  version: v1
paths:
  /test:
    description: Test request
    get:
      summary: Summary
      operationId: getTest
      parameters:
      - deprecated: true
        description: Deprecated parameter
        example: 11-22-33
        in: query
        name: productId
        required: false
        schema:
          type: string
      responses:
        200:
          description: successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  prop1:
                    description: Test prod
                    type: string
```
2. Run this CLI command: `prism mock openapi.yaml`
3. See error:
```
{
    "type": "https://stoplight.io/prism/errors#UNPROCESSABLE_ENTITY",
    "title": "Invalid request body payload",
    "status": 422,
    "detail": "Your request body is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.",
    "validation": [
        {
            "location": [
                "query",
                "productId"
            ],
            "severity": "Warning",
            "code": "deprecated",
            "message": "Query param productId is deprecated"
        }
    ]
}
```

**Expected behavior**
Make a `GET http://localhost:4010/test` request and get a `200 OK` response.

**Environment (remove any that are not applicable):**
 - Library version: 3.1.0

**Additional context**
Removing the `deprecated` flag makes it work correctly.

### PR DESCRIPTION
**Pull request by lag-of-death**
Fixes #714 

## Checklist

- [x] Tests have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## What kind of change does this PR introduce?

_Bug fix_

### CODE REVIEW

File: packages/http/src/mocker/__tests__/HttpMocker.spec.ts [hunk 1]
```diff
@@ -127,7 +127,7 @@ describe('mocker', () => {
       });
     });
 
-    describe('with invalid negotiator response', () => {
```
Comment (XVincentX) [inline packages/http/src/mocker/__tests__/HttpMocker.spec.ts]: Can we also check that it returns a valid response if we have just a warning?

**Comment by XVincentX** [comment]
@lag-of-death Changelog line :)

File: packages/http/src/mocker/__tests__/HttpMocker.spec.ts [hunk 2]
```diff
@@ -127,7 +127,28 @@ describe('mocker', () => {
...
+          }),
+        );
+
+        const mockResult = mock({
+          config: { dynamic: false },
+          resource: mockResource,
+          input: Object.assign({}, mockInput, { validations: [{ severity: DiagnosticSeverity.Warning }] }),
+        })(logger);
+
+        assertRight(mockResult, result => expect(result).toMatchSnapshot());
```
Comment (XVincentX) [inline packages/http/src/mocker/__tests__/HttpMocker.spec.ts]: Since the result is mocked, instead of adding a snapshot — I'd just make sure `negotiateOptionsForValidRequest ` was called and `negotiateOptionsForInvalidRequest ` was NOT called.

File: packages/http/src/mocker/index.ts [hunk 3]
```diff
@@ -89,12 +89,21 @@ function negotiateResponse(
   input: IPrismInput<IHttpRequest>,
   resource: IHttpOperation,
 ) {
-  if (input.validations.length > 0) {
+  const groupedViolations = groupBy(input.validations, (validation) => validation.severity);
```
Comment (StefanDywersant) [inline packages/http/src/mocker/index.ts]: Let's add sugar!

```(karol@MacBook-Pro-Karol prism)$ git diff
diff --git a/packages/http/src/mocker/index.ts b/packages/http/src/mocker/index.ts
index 6f5fd46..2e25686 100644
--- a/packages/http/src/mocker/index.ts
+++ b/packages/http/src/mocker/index.ts
@@ -89,15 +89,18 @@ function negotiateResponse(
   input: IPrismInput<IHttpRequest>,
   resource: IHttpOperation,
 ) {
-  const groupedViolations = groupBy(input.validations, (validation) => validation.severity);
+  const {
+    [DiagnosticSeverity.Error]: errors,
+    [DiagnosticSeverity.Warning]: warnings,
+  } = groupBy(input.validations, (validation) => validation.severity);

-  if (groupedViolations[DiagnosticSeverity.Error]) {
+  if (errors) {
     return handleInputValidation(input, resource);
   } else {
     return pipe(
       withLogger(logger => {
-          if (groupedViolations[DiagnosticSeverity.Warning]) {
-            groupedViolations[DiagnosticSeverity.Warning].forEach((warn) => {
+          if (warnings) {
+            warnings.forEach((warn) => {
               logger.warn({ name: 'VALIDATOR' }, warn.message);
             });
           }```

File: test-harness/specs/deprecated-keyword.only_oas3.txt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+              schema:
+                type: object
+                properties:
+                  prop:
+                    type: string
+                    example: abc
+====server====
+mock -p 4010 ${document}
+====command====
+curl -i http://localhost:4010/test
```
Comment (StefanDywersant) [inline test-harness/specs/deprecated-keyword.only_oas3.txt]: `productId` is not used and it still reports a deprecation warning. I believe it should not report deprecation in such case.

File: CHANGELOG.md [hunk 5]
```diff
@@ -21,6 +21,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - UUIDs are never generated as URNs [#661](https://github.com/stoplightio/prism/pull/661)
 - Core types are now correctly referenced in the HTTP package, restoring the type checks when using the package separately [#701](https://github.com/stoplightio/prism/pull/701)
 - By upgrading Json Schema Faker to the latest version, now the schemas with `additionalProperties:false` / `additionalProperties:true` / `additionalProperties:object` will be correctly handled when dynamic mocking is enabled [#719](https://github.com/stoplightio/prism/pull/719)
+- Making a request to an operation with a `deprecated` parameter is no longer causing Prism to return a 422 response [#731](https://github.com/stoplightio/prism/pull/721)
```
Comment (XVincentX) [inline CHANGELOG.md]:
```suggestion
- Making a request to an operation with a `deprecated` parameter is no longer causing Prism to return a 422 response [#731](https://github.com/stoplightio/prism/pull/731)
```

File: packages/http/src/mocker/index.ts [hunk 6]
```diff
@@ -89,12 +89,24 @@ function negotiateResponse(
...
     return handleInputValidation(input, resource);
   } else {
     return pipe(
-      withLogger(logger =>
-        logger.success({ name: 'VALIDATOR' }, 'The request passed the validation rules. Looking for the best response'),
+      withLogger(logger => {
+          if (warnings) {
+            warnings.forEach((warn) => {
+              logger.warn({ name: 'VALIDATOR' }, warn.message);
+            });
```
Comment (XVincentX) [inline packages/http/src/mocker/index.ts]: You can make this a single line

`warnings.forEach(warn => logger.warn({ name: 'VALIDATOR' }, warn.message))`

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 7]
```diff
@@ -31,6 +31,87 @@ describe('HttpValidator', () => {
...
+              name: "productId"
+            }
+          ],
+        },
+      };
+
+      it('returns warnings', () => {
+        expect(
+          validateInput({
+            // @ts-ignore
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]: Why is this required?

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 8]
```diff
@@ -31,6 +31,87 @@ describe('HttpValidator', () => {
...
+            ],
+            severity: DiagnosticSeverity.Warning
+          }
+        ]);
+      });
+
+      it('does not return warnings', () => {
+        expect(
+          validateInput({
+            // @ts-ignore
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]: ditto

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 9]
```diff
@@ -31,6 +31,87 @@ describe('HttpValidator', () => {
...
+            resource,
+            element: {
+              method: "get",
+              url: {
+                path: "/test",
+                query: {}
+              },
+            },
+          }),
+        ).toEqual([]);
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]:
```suggestion
        ).toHaveLenght(0);
```

File: test-harness/specs/deprecated-keyword.only_oas3.txt (hunk 4 above)
Comment (XVincentX) [reply test-harness/specs/deprecated-keyword.only_oas3.txt]: Testing the deprecation warning is hard unfortunately (because it's a side effect in Prism — and it should not).

When one day we'll introduce the `IO` abstraction from fp-ts this won't be a problem anymore. For now we will have to stick with what we have.

File: packages/http/src/mocker/index.ts [hunk 10]
```diff
@@ -89,12 +89,24 @@ function negotiateResponse(
   input: IPrismInput<IHttpRequest>,
   resource: IHttpOperation,
 ) {
-  if (input.validations.length > 0) {
+  const {
+    [DiagnosticSeverity.Error]: errors,
+    [DiagnosticSeverity.Warning]: warnings,
+  } = groupBy(input.validations, (validation) => validation.severity);
```
Comment (XVincentX) [inline packages/http/src/mocker/index.ts]:
```suggestion
  } = groupBy(input.validations, validation => validation.severity);
```

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 11]
```diff
@@ -31,6 +31,87 @@ describe('HttpValidator', () => {
...
+                mediaType: "application/json",
+                schema: {
+                  type: "object",
+                  properties: {
+                    prop: {
+                      type: "string",
+                      example: "abc"
+                    }
+                  },
+                  $schema: "http://json-schema.org/draft-04/schema#"
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]: You can make this response way shorted, since this is not what we're testing. Provide something basic and save some lines.

File: packages/http/src/validator/__tests__/functional.spec.ts (hunk 7 above)
Comment (lag-of-death) [reply packages/http/src/validator/__tests__/functional.spec.ts]: @XVincentX, this is a mystery:
a_mystery

File: packages/http/src/validator/__tests__/functional.spec.ts (hunk 9 above)
Comment (lag-of-death) [reply packages/http/src/validator/__tests__/functional.spec.ts]: wanted to stay consistent within the file at least https://github.com/stoplightio/prism/blame/fcb5b1f1e1d71769d826e4a531988c8448321799/packages/http/src/validator/__tests__/functional.spec.ts#L29

File: packages/http/src/validator/__tests__/functional.spec.ts (hunk 7 above)
Comment (XVincentX) [reply packages/http/src/validator/__tests__/functional.spec.ts]: Leave it there, I'll figure it out and push a fix.

File: packages/http/src/validator/__tests__/functional.spec.ts (hunk 9 above)
Comment (XVincentX) [reply packages/http/src/validator/__tests__/functional.spec.ts]: Fair.

File: CHANGELOG.md (hunk 5 above)
Comment (lag-of-death) [reply CHANGELOG.md]: actually it's [#721], the link itself was good LOL

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 12]
```diff
@@ -31,6 +31,70 @@ describe('HttpValidator', () => {
...
+            {
+              style: HttpParamStyles.Form,
+              deprecated: true,
+              name: "productId"
+            }
+          ],
+        },
+      };
+
+      it('returns warnings', () => {
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]:
```suggestion
      it('returns a warnings if the property is supplied', () => {
```

File: packages/http/src/validator/__tests__/functional.spec.ts [hunk 13]
```diff
@@ -31,6 +31,70 @@ describe('HttpValidator', () => {
...
+            path: [
+              "query",
+              "productId"
+            ],
+            severity: DiagnosticSeverity.Warning
+          }
+        ]);
+      });
+
+      it('does not return warnings', () => {
```
Comment (XVincentX) [inline packages/http/src/validator/__tests__/functional.spec.ts]:
```suggestion
      it('does not return warnings if the property is not supplied', () => {
```

Comment (lag-of-death) [reply packages/http/src/validator/__tests__/functional.spec.ts]: @XVincentX, ^ this is explained in the `do` portion of the test - https://github.com/stoplightio/prism/pull/721/files/141d4efe845c8ed9f0a888f74f020103c7f806cf#diff-d79142abfe6fcda19641e6dc5c0aebe4R90. `it` only describes what happens

File: packages/http/src/validator/__tests__/functional.spec.ts (hunk 12 above)
Comment (lag-of-death) [reply packages/http/src/validator/__tests__/functional.spec.ts]: https://github.com/stoplightio/prism/pull/721#discussion_r337394390

File: packages/http/src/mocker/index.ts [hunk 14]
```diff
@@ -89,12 +89,20 @@ function negotiateResponse(
...
+  } = groupBy(input.validations, validation => validation.severity);
+
+  if (errors) {
     return handleInputValidation(input, resource);
   } else {
     return pipe(
-      withLogger(logger =>
-        logger.success({ name: 'VALIDATOR' }, 'The request passed the validation rules. Looking for the best response'),
+      withLogger(logger => {
+          warnings && warnings.forEach((warn) => logger.warn({name: 'VALIDATOR'}, warn.message));
```
Comment (StefanDywersant) [inline packages/http/src/mocker/index.ts]:
```suggestion
          warnings && warnings.forEach(warn => logger.warn({name: 'VALIDATOR'}, warn.message));
```

Comment (StefanDywersant) [reply packages/http/src/mocker/index.ts]: Is linter working at all? `warnings && ...` should produce 'assignment or expression expected' error.
